### PR TITLE
CI: split lint-and-tests

### PR DESCRIPTION
**Pull request by evan-gray**
Testing for impoveded consistency / reduced flakiness

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -300,16 +299,31 @@ jobs:
...
-        run: cd node && go test -v -race -ldflags '-extldflags "-Wl,--allow-multiple-definition" ' ./...
       - name: Ensure generated proto matches
         run: |
           rm -rf node/pkg/proto
           docker build --target go-export -f Dockerfile.proto -o type=local,dest=node .
           git diff --name-only --exit-code && echo "✅ Generated proto matches committed proto" || (echo "❌ Generated proto differs from committed proto, run \`rm -rf node/pkg/proto && docker build --target go-export -f Dockerfile.proto -o type=local,dest=node .\` and commit the result" >&2 && exit 1)
 
+  # Run Go tests
+  node-tests:
+    # The tests are slow enough that we want to run it on the self-hosted runner
```
Comment (kcsongor) [inline .github/workflows/build.yml:310]: are they though? we might be able to get away without using the self-hosted one here

Comment (evan-gray) [reply .github/workflows/build.yml:310]: Yes, left this comment in the other PR
https://github.com/wormhole-foundation/wormhole/pull/2278#issuecomment-1399630069
> baseline
>
> - self-hosted
>   - lints 15s
>   - tests 5s
> - cloud
>   - lints 4m 50s
>   - tests 12m 50s

**Comment by evan-gray** [comment]
Going to try pushing up a couple test commits to see if I can get this to fail (like with the tests back where they were)

**Comment by evan-gray** [comment]
Ugh, evidence that the full test fails https://github.com/wormhole-foundation/wormhole/actions/runs/3982537866/jobs/6827087640

**Comment by evan-gray** [comment]
Chatted with @hendrikhofstadt, we'd like to see the lint and tests split out and leave the near tests failing sporadically until we can debug and resolve.

**Comment by evan-gray** [comment]
If this gets consensus, I will switch the required checks.

**Review by jynnantonix — APPROVED** [review]
Is the intention here to split these into two so that the `node-lint` job continues to be required whereas the `node-tests` job becomes optional?

In that case I'd prefer we move just the near tests into a separate job and skip them in the `lint-and-tests` job.  The testing package has a [`SkipNow`](https://pkg.go.dev/testing#T.SkipNow) method we can use for this:

```go
func TestFoo(t *testing.T) {
	if os.Getenv("SKIP_NEAR_TESTS") != "" {
		t.SkipNow()
	}

	// Run the test.
}
```

Then we set the `SKIP_NEAR_TESTS` env variable for the `lint-and-tests` job and add a new `near-tests` job that just does `go test -v -race node/pkg/watchers/near/...`.  This way we keep running the near tests in CI but they don't block unrelated PRs and we continue requiring that all the *other* node tests pass.


If both jobs are going to continue being required then you can ignore everything above.  I don't have any opinions about the lint/test split.  Preemptively approving in this case.

**Review by hendrikhofstadt — APPROVED** [review]

**Comment by evan-gray** [comment]
> Is the intention here to split these into two so that the `node-lint` job continues to be required whereas the `node-tests` job becomes optional?
> 
> In that case I'd prefer we move just the near tests into a separate job and skip them in the `lint-and-tests` job. The testing package has a [`SkipNow`](https://pkg.go.dev/testing#T.SkipNow) method we can use for this:
> 
> ```go
> func TestFoo(t *testing.T) {
> 	if os.Getenv("SKIP_NEAR_TESTS") != "" {
> 		t.SkipNow()
> 	}
> 
> 	// Run the test.
> }
> ```
> 
> Then we set the `SKIP_NEAR_TESTS` env variable for the `lint-and-tests` job and add a new `near-tests` job that just does `go test -v -race node/pkg/watchers/near/...`. This way we keep running the near tests in CI but they don't block unrelated PRs and we continue requiring that all the _other_ node tests pass.
> 
> If both jobs are going to continue being required then you can ignore everything above. I don't have any opinions about the lint/test split. Preemptively approving in this case.

Thanks! I had written up a skip case as well, but we decided against skipping (and hope to just fix them). Will require both tests after this merges 🙂

**Pull request merged**